### PR TITLE
Remove random_fullrank_matrix_distinc_singular_value

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* #67996
* #68523
* #68522
* #69909
* #69908
* #68185
* #68184
* __->__ #68183
* #67014
* #66934
* #67789
* #66933

We do so in favour of
`make_fullrank_matrices_with_distinct_singular_values` as this latter
one not only has an even longer name, but also generates inputs
correctly for them to work with the PR that tests noncontig inputs
latter in this stack.

We also heavily simplified the generation of samples for the SVD, as it was
fairly convoluted and it was not generating the inputs correclty for
the noncontiguous test.

To do the transition, we also needed to fix the following issue, as it was popping 
up in the tests:

Fixes https://github.com/pytorch/pytorch/issues/66856

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D32684853](https://our.internmc.facebook.com/intern/diff/D32684853)

### CODE REVIEW

**Review by mruberry — APPROVED** [review]
LGTM!

Let's wait for the other reviewers.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -4815,93 +4814,33 @@ def sample_inputs_cov(op_info, device, dtype, requires_grad, **kwargs):
...
-        (random_fullrank_matrix_distinct_singular_value(S, dtype=dtype, device=device)[:(S - 2)],
-            lambda usv: (abs(usv[0]), usv[1], abs(slice_V(usv[2])))),  # 'wide_all'
-        (random_fullrank_matrix_distinct_singular_value(S, dtype=dtype, device=device)[:, :(S - 2)],
-            lambda usv: (abs(usv[0][:, :(S - 2)]), usv[1], abs(usv[2]))),  # 'tall_all'
-        (random_fullrank_matrix_distinct_singular_value(S, 2, dtype=dtype, device=device)[..., :(S - 2), :],
-            lambda usv: (abs(usv[0]), usv[1], abs(slice_V(usv[2])))),  # 'wide_all_batched'
-        (random_fullrank_matrix_distinct_singular_value(S, 2, dtype=dtype, device=device)[..., :, :(S - 2)],
-            lambda usv: (abs(usv[0][..., :, :(S - 2)]), usv[1], abs(usv[2]))),  # 'tall_all_batched'
-    )
+def sample_inputs_svd(op_info, device, dtype, requires_grad=False, is_linalg_svd=False, **kwargs):
```
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:4817]: @nikitaved could you have a look at this updated test, see that I did not miss anything. Note that I deleted the `uv_loss`, as the gradients on `U` and `V` are being tested independently, so this should detect that this extra term is correctly implemented.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -10831,7 +10764,8 @@ def ref_pairwise_distance(input1, input2):
     OpInfo('svd',
            op=torch.svd,
            dtypes=floating_and_complex_types(),
-           sample_inputs_func=sample_inputs_svd,
+           sample_inputs_func=partial(sample_inputs_svd, is_linalg_svd=False),
```
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:10767]: Note to self, I should just check the `op_info` inside the generator rather than passing a flag.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:10767]: Either pattern is fine

File: test/test_linalg.py [hunk 3]
```diff
@@ -3198,7 +3199,8 @@ def test_cholesky_solve_out_errors_and_warnings(self, device, dtype):
     @precisionOverride({torch.float32: 2e-3, torch.complex64: 2e-3,
                         torch.float64: 1e-8, torch.complex128: 1e-8})
     def test_inverse(self, device, dtype):
-        from torch.testing._internal.common_utils import random_fullrank_matrix_distinct_singular_value
+        make_fullrank = make_fullrank_matrices_with_distinct_singular_values
+        make_arg = partial(make_fullrank, device=device, dtype=dtype)
```
Comment (nikitaved) [inline test/test_linalg.py:3203]: Do we really care about distinct singular values here?

Comment (nikitaved) [reply test/test_linalg.py:3203]: Nevermind, the condition number it is we care about.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -1228,7 +1227,7 @@ def make_singular_matrix_batch_base(size, rank):
             u_diag_abs_largest = u_diag_abs.max(dim=-1, keepdim=True).values
             u_diag_abs_smallest_idxs = torch.topk(u_diag_abs, k=(n - rank), largest=False).indices
             u.diagonal(0, -2, -1).div_(u_diag_abs_largest)
-            u.diagonal(0, -2, -1)[..., u_diag_abs_smallest_idxs] = torch.finfo(dtype).eps
+            u[..., u_diag_abs_smallest_idxs] = torch.finfo(dtype).eps
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:1230]: This change is fine and does not break anything? I wonder how a 1D index gets properly understood in the 2D "diagonal" context...

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -4815,93 +4814,33 @@ def sample_inputs_cov(op_info, device, dtype, requires_grad, **kwargs):
...
 
-    for a, out_fn in test_cases2:
-        a.requires_grad = requires_grad
-        if is_linalg_svd:
-            kwargs = {'full_matrices': True}
-        else:
-            kwargs = {'some': False}
-        out.append(SampleInput(a, kwargs=kwargs, output_process_fn_grad=out_fn))
+    # The .abs() is to make these functions invariant under multiplication by e^{i\theta}
+    # since the complex SVD is unique up to multiplication of the columsn of U / V a z \in C with \norm{z} = 1
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:4825]: nit: columns

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -4815,93 +4814,33 @@ def sample_inputs_cov(op_info, device, dtype, requires_grad, **kwargs):
...
+    def check_grads(usv):
+        S = usv[1]
+        k = S.shape[-1]
+        U = usv[0][..., :k]
+        Vh = usv[2] if is_linalg_svd else usv[2].mH
+        Vh = Vh[..., :k, :]
+        return (U.abs(), S, Vh.abs())
 
-    return out
+    # We'll set the coorect kwargs in the loop
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:4835]: nit: correct

File: torch/testing/_internal/common_utils.py [hunk 7]
```diff
@@ -2502,8 +2485,15 @@ def make_fullrank_matrices_with_distinct_singular_values(*shape, device, dtype,
...
-        x = (u * s.to(dtype)) @ vh
+        # We choose the singular values to be "around one"
+        # This is to make the matrix well conditioned
+        # s = [2, 3, ..., k+1]
+        s = torch.arange(2, k + 2, dtype=real_dtype, device=device)
+        # s = [2, -3, 4, ..., (-1)^k k+1]
+        s[1::2] *= -1.
+        # 1 + 1/s so that the singular values are in the range [2/3, 3/2]
+        s.reciprocal_().add_(1.)
+        x = (u * s.to(u.dtype)) @ vh
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2496]: nit: maybe we should add a comment explaining that s not being sorted is not an issue here?

File: torch/testing/_internal/common_utils.py [hunk 8]
```diff
@@ -2502,8 +2485,15 @@ def make_fullrank_matrices_with_distinct_singular_values(*shape, device, dtype,
         # TODO: improve the handling of complex tensors here
         real_dtype = t.real.dtype if t.dtype.is_complex else t.dtype
         k = min(shape[-1], shape[-2])
-        s = torch.arange(1., k + 1, dtype=real_dtype, device=device).mul_(1.0 / (k + 1))
-        x = (u * s.to(dtype)) @ vh
+        # We choose the singular values to be "around one"
+        # This is to make the matrix well conditioned
+        # s = [2, 3, ..., k+1]
```
Comment (nikitaved) [inline torch/testing/_internal/common_utils.py:2490]: A mental note: we can start with any value `k > 2` to produce even better conditioned inputs.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:4817]: Is it really tested? The comment says that the output has to depend on both U and V, so I am not sure we do not ignore this special case.

File: torch/testing/_internal/common_utils.py (hunk 8 above)
Comment (Lezcano) [reply torch/testing/_internal/common_utils.py:2490]: imo, `9/4` is a pretty good one and I'd be surprised if it weren't good enough in any of the context we use it in. Good to keep in mind tho.

File: test/test_linalg.py (hunk 3 above)
Comment (Lezcano) [reply test/test_linalg.py:3203]: I'm planning to change the name to `make_wellconditioned` or something along those lines. It'll have the property that the matrix also has distinct singular values, and we'll document that in the function. In practice, it'll just be the function we'll use to generate samples for matrices in `linalg` for most tests

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:1230]: the test was failing and I had to change it. Here all I'm doing is making the whole column(s) to be epsilon hence effectively reducing the norm of the matrix, which is something desirable in this sampling..

File: torch/testing/_internal/common_utils.py (hunk 8 above)
Comment (mruberry) [reply torch/testing/_internal/common_utils.py:2537]: Might as well add a comment

**Comment by Lezcano** [comment]
Addressed the comments. Thanks @nikitaved!

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -4862,93 +4861,34 @@ def sample_inputs_cov(op_info, device, dtype, requires_grad, **kwargs):
...
-            kwargs = {'some': False}
-        out.append(SampleInput(a, kwargs=kwargs, output_process_fn_grad=out_fn))
+    batches = [(), (0, ), (2, )]
+    ns = [0, 2, 5]
 
-    return out
+    # The .abs() is to make these functions invariant under multiplication by e^{i\theta}
+    # since the complex SVD is unique up to multiplication of the columns of U / V a z \in C with \norm{z} = 1
+    # See the docs of torch.linalg.svd for more info
+    def check_grads(usv):
```
Comment (IvanYashchuk) [inline torch/testing/_internal/common_methods_invocations.py:4876]: The deleted function `uv_loss` had a complex return type, while this function has a real-valued return.
I hope with this new function the `imag_term` in https://github.com/pytorch/pytorch/blob/e3bcf64ff84f8e96839e39056b3b90d1bd1f8bbe/torch/csrc/autograd/FunctionsManual.cpp#L2288 is nonzero and its correctness is tested.

**Review by IvanYashchuk — APPROVED** [review]

**Review by nikitaved — APPROVED** [review]
LGTM! Thanks, Mario! Unless there is an issues with the complex term in backward I am happy.

File: torch/testing/_internal/common_methods_invocations.py (hunk 9 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:5001]: It is, indeed, tested, as without that term, complex gradcheck would fail when computing derivatives wrt `U`.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684853).
